File: leetcode/easy/100-sameTree.java
class Solution {
    public boolean isSameTree(TreeNode p, TreeNode q) {
        Stack<TreeNode> stack1 = new Stack<>();
        Stack<TreeNode> stack2 = new Stack<>();
        stack1.push(p);
        stack2.push(q);
        
        while (!stack1.isEmpty() && !stack2.isEmpty()) {
            TreeNode left = stack1.pop();
            TreeNode right = stack2.pop();
            
            if (left == null && right == null) continue;
            if (left == null || right == null) return false;
            if (left.val != right.val) return false;
            
            stack1.push(left.left);
            stack2.push(right.left);
            stack1.push(left.right);
            stack2.push(right.right);
        }
        return stack1.isEmpty() && stack2.isEmpty();
    }
}

File: leetcode/easy/100-sameTree.cpp
// recursive
class Solution {
public:
    bool isSameTree(TreeNode* p, TreeNode* q) {
        if (p == nullptr && q == nullptr) return true;
        if (p == nullptr || q == nullptr) return false;
        if (p->val != q->val) return false;

        return isSameTree(p->left, q->left) && isSameTree(p->right, q->right);
    }
};

// iterative
class Solution {
public:
    bool isSameTree(TreeNode* p, TreeNode* q) {
        stack<pair<TreeNode*, TreeNode*>> stack;
        pair<TreeNode*, TreeNode*> newPair(p, q);
        stack.push(newPair);

        while (!stack.empty()) {
            pair<TreeNode*, TreeNode*> curr = stack.top();
            stack.pop();
        
            if (curr.first == nullptr && curr.second == nullptr) continue;
            if (curr.first == nullptr || curr.second == nullptr) return false;
            if (curr.first->val != curr.second->val) return false;

            pair<TreeNode*, TreeNode*> leftPair(curr.first->left, curr.second->left);
            pair<TreeNode*, TreeNode*> rightPair(curr.first->right, curr.second->right);

            stack.push(leftPair);
            stack.push(rightPair);
        }
        return true;
    }
};
